ActiveAE: Enable display lost callbacks for EGL

popcornmix opened the pull request:
This adds back in support for "Delay after change of refresh rate"
which was lost in the VideoPlayer rework.

popcornmix committed:
ActiveAE: Enable display lost callbacks for EGL

This adds back in support for "Delay after change of refresh rate"
which was lost in the VideoPlayer rework.

popcornmix commented:
Ping @FernetMenta

FernetMenta commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -291,7 +291,7 @@ CActiveAE::~CActiveAE()
 void CActiveAE::Dispose()
 {
-#if defined(HAS_GLX) || defined(TARGET_DARWIN)
+#if defined(HAS_GLX) || defined(HAS_EGL) || defined(TARGET_DARWIN)
```
the entire ifdeffery can be dropped. Windowing has to support it

FernetMenta commented:
+1

popcornmix commented:
Removed ifdefs.
jenkins build this please

FernetMenta commented:
> Removed ifdefs.

forgot to push?

popcornmix commented:
Yes. Pushed now.

popcornmix commented:
jenkins build this please

popcornmix commented:
Seems this causes a hang in some circumstances so don't merge yet.

popcornmix commented:
Added a fix for the hang discovered.
jenkins build this please

FernetMenta commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -1294,6 +1294,7 @@ void CVideoPlayer::Process()
       CSingleLock lock(m_StateSection);
       if (m_displayLost)
       {
+        CSingleExit unlock(m_StateSection);
```
CSingleExit makes no sense here. Why did you ignore my advice to drop the lock completely?

popcornmix replied:
Not sure why it makes no sense. Seems a valid fix that is done similarly in many places in the code.
But I can switch to using atomics for m_displayLost if that is desired.

FernetMenta replied:
You would use CSingleExit if you want to exit a lock and grab it again. This is not the case here. It would grab the lock again and on the next line the outer lock gets destructed anyway.

FernetMenta replied:
look at the entire block:
```

    // check display lost
    {
      CSingleLock lock(m_StateSection);
      if (m_displayLost)
      {
        CSingleExit unlock(m_StateSection);
        Sleep(50);
        continue;
      }
    }
```

popcornmix commented:
Updated the PR to pause/resume video playback when display is lost. This works a lot better for me, especially when "Delay after change of refresh rate" is used.
@FernetMenta ?

FernetMenta commented:
I already mentioned that the entire WinSystemEGL will disappear. It would be great if you could start factoring out Pi from there. EGL is not responsible for creating or destroying windows, it is just the glue code to the GL context.

popcornmix committed:
rendering: Always call Flip but don't sleep when playing video

This allows delay after refresh rate change to continue when video is playing with no gui overlay

popcornmix commented:
@FernetMenta 
This is updated to include the fix for stalling when video is rendered to a separate layer.

FernetMenta commented:
jenkins build this please

FernetMenta commented:
@popcornmix build errors on Linux and Windows

popcornmix commented:
jenkins build this please

popcornmix commented:
@FernetMenta jenkins is happy, okay to merge?
